File: 508. Most Frequent Subtree Sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    // dfs: O(N)time & space, 53.98%, ok
    vector<int> findFrequentTreeSum(TreeNode* root) {
        vector<int> ret;
        unordered_map<int, int> cnt;
        int maxCnt = -1;
        int rootSum = dfs(root, cnt, maxCnt);
        for(auto pair: cnt){
            if(pair.second == maxCnt)
                ret.push_back(pair.first);
        }
        return ret;
    }

    int dfs(TreeNode* root, unordered_map<int, int>& cnt, int& maxCnt){
        if(!root) return 0;
        int currSum = root->val + dfs(root->left, cnt, maxCnt) + dfs(root->right, cnt, maxCnt);
        cnt[currSum]++;
        maxCnt = max(maxCnt, cnt[currSum]);
        return currSum;
    }


};

File: 508. Most Frequent Subtree Sum.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    
    # dfs & hashmap: O(N) time & space
    def findFrequentTreeSum(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        def treeSum(node, cnt):
            if not node:
                return 0
            ret = treeSum(node.left, cnt) + treeSum(node.right, cnt) + node.val
            cnt[ret] = cnt.get(ret, 0) + 1
            return ret
            
        cnt = {}
        treeSum(root, cnt)
        maxCnt = -1
        for num, count in cnt.items():
            if count > maxCnt:
                maxCnt = count
        return [num for num in cnt if cnt[num] == maxCnt]
        
        